### PR TITLE
docs: make `HeroService#gerHero()` in `toh-pt5` compatible with TS strict mode

### LINKED ISSUE #41207: TypeScript not happy with some example code in tutorial (Tour of Heros) part 5 (navigation)
# 📚 Docs or angular.io bug report

### Description

In tutorial Tour of Heros part 5 (add navigation), some example codes are failing typescript checks. 
Minor issues, but still better to be fixed some time.

### What's the affected URL?**
https://angular.io/tutorial/toh-pt5

### cleaned up version of code
In src/app/heroes/heroes.component.ts (cleaned up):
image
my quick fix: `heroes: Hero[] = [];` 

### getHero function in HeroService
https://angular.io/tutorial/toh-pt5#add-heroservicegethero
image
image
suggestion: add a fallback when no id matches. 
```
getHero(id: number): Observable<Hero> {
    this.messageService.add(`HeroService: fetched hero id=${id}`);
    let result = HEROES.find(hero => hero.id === id);
    return of(!!result ? result: {id: id, name: "No result"});
  }
```

### PR DESCRIPTION
**Pull request by gkalpak**
☝️ 

Fixes #41207.

### CODE REVIEW

File: aio/content/examples/toh-pt5/src/app/hero.service.ts [hunk 1]
```diff
@@ -12,16 +12,21 @@ export class HeroService {
...
 
   // #docregion getHero
   getHero(id: number): Observable<Hero> {
-    // TODO: send the message _after_ fetching the hero
+    const hero = HEROES.find(h => h.id === id);
+
+    if (hero === undefined) {
+      throw new Error(`Failed to fetch hero id=${id}`);
+    }
+
```
Comment (AndrewKushnir) [inline aio/content/examples/toh-pt5/src/app/hero.service.ts:27]: Just curious if we can use a non-null assertion statement to keep things simple in the example code:
```suggestion
    const hero = HEROES.find(hero => hero.id === id)!;
```
If we throw an error in a service, I guess we should also illustrate how to handle that, which goes beyond simple examples.

Comment (gkalpak) [reply aio/content/examples/toh-pt5/src/app/hero.service.ts:27]: I switched so many times between the two options while working on this locally :grin:
Neither option is ideal for the tutorial (non-null assertion is too magic and covers up cases where the ID does not exist and throwing errors without handling them offers a bad UX).

This code is [enhanced in the next chapter][1] of the tutorial to use `HttpClient` for fetching the hero and handling errors, so I don't think it is worth introducing error handling here. How about adding an `as Hero` assertion and add a `TODO` comment to handle the case where the hero cannot be found?

[1]: https://github.com/angular/angular/blob/7d5b60e4097195168c454c18d1b9352ea8c9b9a4/aio/content/examples/toh-pt6/src/app/hero.service.ts#L39-L49

Comment (gkalpak) [reply aio/content/examples/toh-pt5/src/app/hero.service.ts]: I've pushed a fixup commit. WDYT?

File: aio/content/examples/toh-pt5/src/app/hero.service.ts [hunk 2]
```diff
@@ -12,16 +12,17 @@ export class HeroService {
...
     this.messageService.add('HeroService: fetched heroes');
-    return of(HEROES);
+    return heroes;
   }
 
   // #docregion getHero
   getHero(id: number): Observable<Hero> {
-    // TODO: send the message _after_ fetching the hero
+    // TODO: Handle failing to find a hero with the specified `id`.
+    const hero = HEROES.find(h => h.id === id) as Hero;
```
Comment (AndrewKushnir) [inline aio/content/examples/toh-pt5/src/app/hero.service.ts:23]: May be we can update the comment to something like this?
```suggestion
    // In order to keep the example app simple, we assume that
    // a Hero with the provided id is always present in the list.
    const hero = HEROES.find(hero => hero.id === id) as Hero;
```

**Review by AndrewKushnir — APPROVED** [review]
LGTM 👍 Just added a proposal on rephrasing the comments.

File: aio/content/examples/toh-pt5/src/app/hero.service.ts (hunk 1 above)
Comment (AndrewKushnir) [reply aio/content/examples/toh-pt5/src/app/hero.service.ts]: > I've pushed a fixup commit. WDYT?

Looks great, just added a note with a proposal on rephrasing comments.

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: docs-getting-started-and-tutorial

**Pull request closed without merging**